src/hb-ot-layout-gsubgpos.hh

```diff
@@ -3678,57 +3678,60 @@ struct GSUBGPOS
                                 const hb_set_t *feature_indices,
                                 hb_map_t *duplicate_feature_map /* OUT */) const
   {
+    hb_hashmap_t<hb_tag_t, hb_set_t *, (unsigned)-1, nullptr> tag_features;
```

Comment:
Since feature indices are sorted by tag's alphabetically that means all feature with the same tag are consecutive. So you don't need to store the set for each tag. Rather you could have one set, which you clear every time you encounter a feature with a different tag than the previous.